iplist.8
.TH iplist 8 "2008-10-14" "0.30" "iplist.sf.net"
.SH NAME
iplist - List based packet handler
.SH SYNOPSIS
.B iplist
[\fIoptions\fP] [\fIfile\fP...]
.br
.B iplist
\fB-\-show\fR
.br
.B iplist
\fB-\-kill\fR
.br
.B iplist
\fB-\-delete -\-queue\-num\fR=\fI0\-65535\fR
.SH DESCRIPTION
\fBiplist\fP is a list based packet handler which uses 
the netfilter netlink-queue library (kernel 2.6.14 or 
later). It filters by IP-address and is optimized for 
thousands of IP-address ranges.
.SH OPTIONS
.TP
\fB\-p\fR, \fB\-\-policy\fR=\fBaccept\fR|\fBdrop\fR|\fBrepeat\fR
default action for packets. This is like the iptables policy 
setting. Packets that are repeated are send back to the source 
chain (IN-, OUTPUT or FORWARD). Default is accept.
.TP
\fB\-P\fR, \fB\-\-policy-mark\fR=\fIvalue[/mask]\fR
mark packets with a non-zero value. 
Values can be in dec, hex (prefix 0x or 0X) or oct (prefix 0).
If mask is specified value will be logically ANDed with mask.
This option is intended to be used with a repeat policy. 
Default is 0 (don't mark).
.TP
\fB\-t\fR, \fB\-\-target\fR=\fBaccept\fR|\fBdrop\fR|\fBrepeat\fR
target for packets that match rule. This option allows to specify 
how packets that match rules are handled. If this option is used it
overrides the target specified in dat or ipl files. Default is drop.
.TP
\fB\-T\fR, \fB\-\-target-mark\fR=\fIvalue[/mask]\fR
mark packets with a value. Same as the policy-mark option but for 
packets that match rules.  Default is 0 (don't mark).
.TP
\fB\-n\fR, \fB\-\-queue\-num\fR=\fI0\-65535\fR
iptables queue number. There is a same option for iptables. 
Default is 0.
.TP
\fB\-i\fR, \fB\-\-insert\fR
insert a new queue. This option allows to use more than one
queue and to have different files with different targets 
associated with one queue.
.TP
\fB\-d\fR, \fB\-\-delete\fR
delete an existing queue by number 
.TP
\fB\-s\fR, \fB\-\-show\fR
show current queues and exit
.TP
\fB\-\-strict\-ip\fR
IP-addresses can be in hex, oct and dec. E.g: 192.0xff.010
would be an allowed IP-address where the first part is dec, 
the second is hex, the third is oct. Without this option
all IP-addresses must be in decimal notation.
.TP
\fB\-r\fR, \fB\-\-pid\-file\fR=\fIfile\fR
specify PID file location. Default is /var/run/iplist.pid when
the uid is root otherwise /tmp/iplist.pid is used.
.TP
\fB\-\-stdout\fR
log to terminal. Same as --log-file=/dev/stdout.
.TP
\fB\-f\fR, \fB\-\-log\-file\fR=\fIfile\fR
network traffic related output. This option has no effect, if the 
loglevel option isn't used or has the value none.
.TP
\fB\-l\fR, \fB\-\-log\-level\fR=\fBall\fR|\fBmatch\fR|\fBnone\fR
specify network traffic loglevel. If the verbose option is used, 
it will print detailed packet information. Default is none.
.TP
\fB\-o\fR, \fB\-\-output\fR=\fIfile\fR
convert files to another format and exit. The output will be
gzip compressed. 
.TP
\fB\-O\fR, \fB\-\-output\-fmt\fR=\fBipl\fR|\fBdat\fR|\fBp2p\fR
specify output format. Default is ipl binary format.
.TP
\fB\-d\fR, \fB\-\-daemon\fR
start as daemon in background. Note that only one instance of 
iplist is possible.
.TP
\fB\-k\fR, \fB\-\-kill\fR
kill running iplist instance and exit. This option kills iplist 
whether it runs as a daemon or in foreground.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
increase verbosity. Shows detail information about the program and the 
network activities. 
.TP
\fB\-q\fR, \fB\-\-quiet\fR
suppress non-error messages. iplist uses syslog. If this 
option is used only errors will be logged to syslog.
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
output version information and exit
.PP
Supported file formats: p2p, dat, csv and ipl. Use "-" to read from stdin.
Files can optionally be compressed with gzip.
.SH ENVIRONMENT
.TP
IPLIST_LISTDIR
If IPLIST_LISTDIR is set, it specifies a directory to look for list
files. 
.SH EXAMPLES
.TP
1. Using repeat policy
The policies ACCEPT and DROP are terminating the traversal 
of packets through the firewall. Sometimes however it is 
desired to let the packet through iplist so that other
iptables rules can be applied. This is achieved by marking
and repeating packets. This is an often used method in IPSEC-setups.

The idea is to send packets that are not marked with a policy-mark 
to iplist. If a packet doesn't match any IP-addresses, iplist
marks and repeats it to the source chain.

This rule would send every packet which isn't marked with a policy-mark
to iplist

#> iptables -I INPUT -m mark ! --mark 1 -j NFQUEUE

iplist would then be started like this

#> iplist --policy repeat --policy-mark 1 --target drop file.p2p

The result of this setup is that any incoming packets that match 
IP-addresses in file.p2p are dropped, everything else is handled by the 
rest of the iptables configuration.
.TP
2. Blocking specific P2P traffic 
Let's say a P2P application listens on ports 6991 - 6999.
The goal is not to allow connections with all clients from 
a bad.p2p file. Inserting the following rule into iptables
would block in- and outcoming packets that attempt to establish
a new conncection.

#> iptables -I INPUT -p tcp -m state --state NEW --dport 6991:6999 -j NFQUEUE
.br
#> iptables -I OUTPUT -p tcp -m state --state NEW --sport 6991:6999 -j NFQUEUE

Then iplist can be started like this

#> iplist -p accept -t drop --stdout -l match bad.p2p

All packets that are dropped are logged and printed to the terminal.
.TP
3. Rejecting Packtes
Rejecting packets is done using the repeat option like in example 1.
Target matches are marked with target-mark and repeated. The result
is that the iptables rule with "--mark <target-mark>" can be applied.

#> iptables -I INPUT 1 -m mark --mark 2 -j REJECT
.br
#> iptables -I INPUT 2 -m mark ! --mark 1 -j NFQUEUE

#> iplist -p repeat -P 1 -t repeat --target-mark 2 bad.p2p

The order of the iptables rules is important. The reject rule must be
first.
.SH AUTHOR
Written by Serkan Sakar.
.SH "REPORTING BUGS"
Report bugs to <uljanow@users.sourceforge.net>
or on https://sourceforge.net/tracker/?atid=966442&group_id=198679
.SH COPYRIGHT
Copyright \(co 2010 Serkan Sakar <uljanow@users.sourceforge.net>
.br
This is free software.  You may redistribute copies of it under the terms 
of the GNU General Public License <http://www.gnu.org/licenses/gpl.html>.
There is NO WARRANTY, to the extent permitted by law.



changelog
2013-04-13  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.30)
	* fixed compiler issues 
	* fixed thread return value bugs

2010-10-17  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.29)
	* fixed compiler issues 
	* faulty lines don't cause an exit
	* fixed broken file names during update
	* fixed lintian errors (deb packages)

2010-03-11  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.28)
	* fixed debian ifup script
	* added GUI watchdog
	* Improved GUI log routines
	* Reduced GUI memory usage
	* added IP counter
	* improved reload behaviour of list manager

2009-10-07  Serkan Sakar <uljanow@users.sourceforge.net>
	
	* release (0.27)
	* improved performace of list routines
	* fixed theme of context menu in whitelist manager
	* improved status check routines
	* changed UI to include a log-level combobox
	* added info dialog if log entry is double-clicked
	* replaced tail dependency
	* removed GUI_LAST_LOG_LINES option

2009-05-01  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.26)
	* fixed broken LESS_MEMORY option
	* fixed regex bug (2683790) due to latest grep 2.5.4
	* fixed if-up script warning due to exit code
	* fixed autostart bug in ubuntu jaunty
	* added theme option to GUI
	* improved kde support
	* added generic su-to-root wrapper to ipblock
	* removed dependencies on desktop-environments

2009-02-19  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.25)
	* fixed gzip error message
	* moved icon to /usr/share/pixmaps/
	* removed "-u root" parameter in desktop file
	* added gentoo init script
	* fixed linux mint menu starter issue
	* fixed OpenSuse 11.1 rpm-lint errors
	* fixed format of range names in logfile
	* added if-up script (Debian)
	* fixed GUI problem in Arch Linux
	* whitelisted ranges allow incoming traffic (default setting)
	* added -p option to restart iplist only (used by ipblock.cron)
	* removed update IP ranges in allow.p2p due to default http whitelist

2008-11-30  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.24)
	* enabled busyIcon of the GUI
	* fixed compiler warnings 
	* fixed lintian error due to force-reload
	* added IGN_PROTO_* options to ignore protocols
	* changed Debian java dependency to java6-runtime
	* fixed opensuse autostart issues
	* added new whitelist manager
	* added context menu to allow logged IPs
	* added LESS_MEMORY option and compiler flag
	* renamed option to GUI_START_HIDDEN
	* added GUI_THEME option
	* added GUI_WHITELIST_* options
	* added log column Type
	* added -i parameter to ipblock
	* added automatical restart to list manager
	* added online help menuitem

2008-10-18  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.23)
	* new GUI with native look & feel in GNOME
	* added systray support (only Metacity currently)
	* added new ipblock icon
	* added GUI logfile 
	* added GUI_START_MINIMIZED option
	* added GUI_LAST_LOG_LINES option
	* added GUI_AUTOSCROLL option
	* http and domain ports are enabled in OUTPUT by default
	* added Mandriva Linux 2009 support

2008-10-04  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.22)
	* added BLOCK_TARGET_* options and set the default 
	  action for incoming blocked packets to DROP
	* modified GUI log entries
	* fixed new GUI logging bug

2008-09-14  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.21)
	* added BLOCK_LIST_*, ALLOW_LIST_* and 
	  IPTABLES_CHAIN_ALLOW options
	* modified GUI to show the last 12 log entries
	* allow lists can optionally be compressed with gzip
	* extended LOG_IPTABLES feature to show allowed packets
	* Update messages are appended to logfile
	* added Update dialog for adding lists from URL 
	* default allow list is only valid in the output chain
	* added more countries (e.g. russia, india) 
	* improved GUI cleanup

2008-07-29  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.20)
	* fixed gcc-4.3 compiling issues
	* added menu starter for fedora
	* moved README.lists to /etc/ipblock.lists
	* added import URL feature in GUI
	* added OpenSuSE 11 support

2008-01-25  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.19)
	* removed IANA Private IP addresses from allow.p2p

2007-12-30  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.18)
	* made init scripts more distribution specific 
	* added ALLOW_LIST option
	* changed URL list format
	* integrated country lists

2007-12-22  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.17)
	* fixed bug 1852759 (thanks to ekimregnirps)
	* improved cleanup of iptables rules
	* resolved conflict between using GUI and CLI
	* added cron job to auto-update lists
	* added AUTOSTART option and init script
	* minor GUI layout changes

2007-12-8  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.16)
	* minor GUI bugfixes 
	* improved robustness of IPblock
	* starting IPblock takes less time
	* changed max files per queue to 32
	* iplist doesn't terminate on unsupported file types
	* minor changes to logging
	* added IPTABLES_CHAIN option
	* added IPTABLES_LOG option
	* added http_proxy option
	* added DEBUG switch

2007-10-11  Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.15.1)
	* fixed suspend error
	* fixed kernel 2.6.23 support

2007-08-29	Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.15)
	* fixed custom ports bug
	* fixed wrong error message on successful start (ipblock)
	* added "add URL" functionality

2007-08-20	Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.14)
	* added quiet option
	* added support for value/mask marks
	* added support for PRE- and POSTROUTING chains
	* initial release of ipblockUI
	* fixed compatibility issues due to mark values

2007-07-13	Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.13)
	* renamed nfqueue to iplist
	* added output-fmt option
	* ipblock.sh can ignore ports 

2007-07-10	Serkan Sakar <uljanow@users.sourceforge.net>

	* release (0.12)
	* improved reading/writing routines

src/iplist.cc
/*
	iplist - List based packet handler
	Copyright (C) 2010 Serkan Sakar <uljanow@users.sourceforge.net>

	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation; either version 2 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  
	02110-1301, USA
*/

#include <iostream>
#include <iomanip>
#include <stdexcept>
#include <map>
#include <fstream>
#include <cerrno>
#include <sstream>
#include <cstdlib>
#include <cerrno>
#include <string.h>

#include <regex.h>
#include <getopt.h>
#include <syslog.h>
#include <fcntl.h>
#include <sys/resource.h>
#include <sys/types.h> 
#include <sys/stat.h> 

#include "iplist.h"
#include "list.h"
#include "log.h"

volatile std::sig_atomic_t iplist::quit = 0;
int iplist::daemon_flag = 0;
int iplist::verbose_flag = 0;
int iplist::quiet_flag = 0;

sigset_t iplist::mask;
pthread_mutex_t iplist::lock = PTHREAD_MUTEX_INITIALIZER;
pthread_cond_t iplist::wait = PTHREAD_COND_INITIALIZER;

iplist::process iplist::ps;

const char* LISTDIR = "IPLIST_LISTDIR";
const mode_t MODE = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH;

static std::ostream& print_version(std::ostream& os)
{
	return os << "iplist " << VERSION << 
		"\nCopyright (C) 2010 Serkan Sakar <uljanow@users.sourceforge.net>" 
		<< std::endl;
}

static void print_usage(std::ostream& os, int exit_status)
{
	int width = 35;
 	os.setf(std::ios::left);
	print_version(os) <<
		"\nUsage: \n\n " <<
		"iplist [options] [<file>...]\n " << 
		"iplist --show\n " << 
		"iplist --kill\n " << 
		"iplist --delete --queue-num=<0-65535>\n " << 
		"\nOptions: \n\n" << std::setw(width) << 
		" -p, --policy=accept|drop|repeat" << "default action for packets\n" << std::setw(width) << 
		" -P, --policy-mark=<value[/mask]>" << "mark packets with a non-zero value\n" << std::setw(width) << 
		" -t, --target=accept|drop|repeat" << "target for packets that match rule\n" << std::setw(width) << 
		" -T, --target-mark=<value[/mask]>" << "mark packets with a non-zero value\n\n" << std::setw(width) << 

		" -n, --queue-num=<0-65535>" << "iptables queue number\n" << std::setw(width) << 
		" -i, --insert" << "insert a new queue\n" << std::setw(width) << 
		" -d, --delete" << "delete an existing queue by number\n" << std::setw(width) << 
		" -s, --show" << "show current queues and exit\n\n" << std::setw(width) << 

		"     --stdout" << "log to terminal\n" << std::setw(width) << 
		" -f, --log-file=<file>" << "network traffic related output\n" << std::setw(width) << 
		" -l, --log-level=all|match|none" << "specify network traffic log level \n\n" << std::setw(width) << 

		"     --strict-ip" << "IP-addresses can be in hex, oct and dec\n" << std::setw(width) << 
		" -r, --pid-file=<file>" << "specify PID file location\n" << std::setw(width) << 
		" -o, --output=<file>" << "convert files to another format and exit\n" << std::setw(width) << 
		" -O, --output-fmt=ipl|dat|p2p" << "specify output format\n" << std::setw(width) << 
		" -b, --daemon" << "start as daemon in background\n" << std::setw(width) << 
		" -k, --kill" << "kill running iplist instance and exit\n" << std::setw(width) << 
		" -v, --verbose" << "increase verbosity\n" << std::setw(width) << 
		" -q, --quiet" << "suppress non-error messages\n" << std::setw(width) << 
		" -h, --help" << "display this help and exit\n" << std::setw(width) <<
		" -V, --version" << "output version information and exit\n" << std::setw(width) <<
		"\nSupported file formats are p2p, dat, csv and ipl. Use \"-\" to read from stdin."
		"\nFiles can optionally be compressed with gzip."  << std::endl;
	exit(exit_status);
}

iplist::process::process():
	pid_fd(-1), key(0), pid(0), msqid(0),
	pid_file("/var/run/iplist.pid"), fifo("/tmp/.iplist")
{
	if (mkfifo(fifo.c_str(), S_IRUSR | S_IWUSR) == -1)
		if (errno != EEXIST)
			throw std::runtime_error("can't create fifo");
}

iplist::process::~process()
{
	if (pid) return; 

	if (!lock_test(pid_file, F_WRLCK, 0, 1)) {
		close(pid_fd);
		unlink(pid_file.c_str());
		unlink(fifo.c_str());
		msgctl(msqid, IPC_RMID, NULL);
	}
}

void iplist::process::daemonize()
{
	int fd0, fd1, fd2;
	rlimit rl;

 	umask(0);

	if (getrlimit(RLIMIT_NOFILE, &rl) < 0)
		throw std::runtime_error("can't get file descriptor limit");

	if ((pid = fork()) < 0)
		throw std::runtime_error("can't fork");
	else if (pid)
		exit(EXIT_SUCCESS);

	setsid();

	if (std::signal(SIGHUP, SIG_IGN) == SIG_ERR)
		throw std::runtime_error("can't ignore SIGHUB");

	if ((pid = fork()) < 0)
		throw std::runtime_error("can't fork");
    else if (pid)
		exit(EXIT_SUCCESS);

	if (chdir("/") < 0)
		throw std::runtime_error("can't change directory to /");

	if (rl.rlim_max == RLIM_INFINITY) 
		rl.rlim_max = 1024;
	for (u_int i = 0; i < rl.rlim_max; i++) close(i);

	fd0 = open("/dev/null", O_RDWR);
	fd1 = dup(0);
	fd2 = dup(0);

	if (fd0 != 0 || fd1 != 1 || fd2 != 2)
		throw std::runtime_error("unexpected file descriptors");
}

bool iplist::process::lock_file(const std::string& file, int cmd,
		short l_type, off_t l_start, off_t l_len)
{
	int fd;
	flock fl;
	fl.l_type = l_type;
	fl.l_whence = SEEK_SET;
	fl.l_start = l_start;
	fl.l_len = l_len;

	if ((fd = open(file.c_str(), O_RDWR | O_CREAT, MODE)) == -1)
		throw std::ios_base::failure("can't open " + file);

	if (fcntl(fd, cmd, &fl) == -1) {
		if (errno == EACCES || errno == EAGAIN)
			return false;
		else
			throw std::ios_base::failure("can't lock " + file);
	}

	return true;
}

void iplist::process::write_pid()
{
	std::ostringstream is;
	is << (long)getpid() << "\n";

	if ((pid_fd = open(pid_file.c_str(), O_RDWR | O_CREAT, MODE)) == -1)
		throw std::ios_base::failure("can't open " + pid_file);

	if (ftruncate(pid_fd, 0))
		throw std::ios_base::failure(strerror(errno));

	if (!write(pid_fd, is.str().c_str(), is.str().size())) 
		throw std::ios_base::failure("can't write PID to " + pid_file);
}

pid_t iplist::process::lock_test(const std::string& file, short l_type,
		off_t l_start, off_t l_len)
{
	int fd;
	flock fl;
	fl.l_type = l_type;
	fl.l_whence = SEEK_SET;
	fl.l_start = l_start;
	fl.l_len = l_len;

	if ((fd = open(file.c_str(), O_RDONLY, MODE)) == -1)
		return 0;

    if (fcntl(fd, F_GETLK, &fl) == -1)
		throw std::ios_base::failure("can't get lock of " + file);

    return (fl.l_type == F_UNLCK) ? 0 : fl.l_pid;
}

void iplist::process::create_msq(key_t k)
{
	msqid = msgget((key = k), IPC_CREAT | 0600);
}

void iplist::parse_cmdline(int argc, char** argv)
{ 
	const char* short_opt = "p:P:t:T:n:idsr:f:l:o:O:bkvqhV";	
	const option long_opt[] = {
		{"policy", required_argument, NULL, 'p'},
		{"policy-mark", required_argument, NULL, 'P'},
		{"target", required_argument, NULL, 't'},
		{"target-mark", required_argument, NULL, 'T'},
		{"queue-num", required_argument, NULL, 'n'},
		{"insert", no_argument, NULL, 'i'},
		{"delete", no_argument, NULL, 'd'},
		{"show", no_argument, NULL, 's'},
		{"pid-file", required_argument, NULL, 'r'},
		{"stdout", no_argument, NULL, '1'},
		{"log-file", required_argument, NULL, 'f'},
		{"log-level", required_argument, NULL, 'l'},
		{"strict-ip", no_argument, &list::strict_ip, 1},
		{"output", required_argument, NULL, 'o'},
		{"output-fmt", required_argument, NULL, 'O'},
		{"daemon", no_argument, &iplist::daemon_flag, 1},
		{"kill", no_argument, NULL, 'k'},
		{"verbose", no_argument, &iplist::verbose_flag, 1},
		{"quiet", no_argument, &iplist::quiet_flag, 1},
		{"help", no_argument, NULL, 'h'},
		{"version", no_argument, NULL, 'V'},
		{NULL, 0, NULL, 0}
	};
	bool client = false;
	std::pair<std::string, list::file_t> output("", list::IPL);
	int opt;
	int8_t target = -1;
	job_ptr j(new job);
	const size_t BUF_SIZE = 100;
	std::auto_ptr<char> buf(new char[BUF_SIZE]);

	if (!getcwd(buf.get(), BUF_SIZE))
		throw std::runtime_error("can't get working directory");

	std::string wd(buf.get());
	wd += "/";

	while ((opt = getopt_long(argc, argv, short_opt, long_opt, NULL)) != -1) {
		switch (opt) {
		case 'p':
			if (!strcasecmp(optarg, "repeat"))
				j->policy = NF_REPEAT;
			else if (!strcasecmp(optarg, "accept"))
				j->policy = NF_ACCEPT;
			else if (!strcasecmp(optarg, "drop"))
				j->policy = NF_DROP;
			else if (!strcasecmp(optarg, "queue"))
				j->policy = NF_QUEUE;
			else
				throw std::invalid_argument("invalid policy argument " 
						+ std::string(optarg));
			break;
		case 'P': 
			strncpy(buf.get(), optarg, BUF_SIZE);
			j->policy_mark = strtoul(strtok(buf.get(), "/"), NULL, 0); 

			if (char* mask = strtok(NULL, "/"))	
				j->policy_mark &= strtoul(mask, NULL, 0); 
			break;
		case 't':
			if (!strcasecmp(optarg, "drop"))
				target = NF_DROP;
			else if (!strcasecmp(optarg, "accept"))
				target = NF_ACCEPT;
			else if (!strcasecmp(optarg, "repeat"))
				target = NF_REPEAT;
			else if (!strcasecmp(optarg, "queue"))
				target = NF_QUEUE;
			else
				throw std::invalid_argument("invalid target argument " 
						+ std::string(optarg));
			break;
		case 'T':
			strncpy(buf.get(), optarg, BUF_SIZE);
			j->target_mark = strtoul(strtok(buf.get(), "/"), NULL, 0); 

			if (char* mask = strtok(NULL, "/"))	
				j->target_mark &= strtoul(mask, NULL, 0); 
			break;
		case 'n':
			j->nfq_num = strtoul(optarg, NULL, 0);
			break;
		case 'i':
			if (!ps.lock_test(ps.pid_file, F_WRLCK, 0, 1)) 
				daemon_flag = 1;
			else
				client = true;
			break;
		case 'd':
			j->req = DELETE;
			client = true;
			break;
		case 's':
			j->req = WRITE;
			client = true;
			j->pid = getpid();
			break;
		case '1':
			log::logfile = "/dev/stdout"; 
			break;
		case 'f':
			log::logfile = (!strncmp(optarg, "/", 1)) ? optarg : wd + optarg; 
			break;
		case 'l':
			if (!strcasecmp(optarg, "none"))
				log::loglevel |= LOG_NONE;
			else if (!strcasecmp(optarg, "match"))
				log::loglevel |= LOG_MATCH;
			else if (!strcasecmp(optarg, "all"))
				log::loglevel |= LOG_ALL;
			else
				throw std::invalid_argument("invalid loglevel argument "
						+ std::string(optarg));
			break;
		case 'o':
			output.first = optarg;
			break;
		case 'O':
			if (!strcasecmp(optarg, "ipl"))
				output.second = list::IPL;
			else if (!strcasecmp(optarg, "dat"))
				output.second = list::DAT;
			else if (!strcasecmp(optarg, "p2p"))
				output.second = list::P2P;
			else
				throw std::invalid_argument("invalid output format "
						+ std::string(optarg));
			break;
		case 'b':
			daemon_flag = 1; 
			break;
		case 'k': 
			if (pid_t pid = ps.lock_test(ps.pid_file, F_WRLCK, 0, 1))
				exit(kill(pid, SIGTERM));

			throw std::runtime_error("no running iplist instance found");
		case 'r':
			ps.pid_file = (!strncmp(optarg, "/", 1)) ? optarg : wd + optarg; 
			break;
		case 'v':
			verbose_flag = 1; 
			break;
		case 'q':
			quiet_flag = 1; 
			break;
		case 'V': 
			print_version(std::cout);
			exit(EXIT_SUCCESS);
		case 0: break;
		case 'h':
			print_usage(std::cout, EXIT_SUCCESS);
		case '?':
		default:
			print_usage(std::cerr, EXIT_FAILURE);
		}
	}
	if (verbose_flag) {
		setlogmask(LOG_UPTO(LOG_DEBUG));
		log::loglevel |= LOG_VERBOSE;
	} else if (quiet_flag) 
		setlogmask(LOG_UPTO(LOG_ERR));

	if (log::logfile.empty()) 
		log::loglevel = LOG_NONE;

	if (!argv[optind] && j->req == INSERT)
		j->req = NONE;

	if (argv[optind]) {
		if (!strncmp(argv[optind], "-", 1))
			if (client) {
				j->pid = getpid();
				j->add_file(ps.fifo.c_str(), target);
			} else
				j->add_file("/dev/stdin", target);
		else {
			std::string listd;
			
			if (char* env = getenv(LISTDIR))
				listd = env + std::string("/");

			for (u_int i = optind; argv[i]; i++) {
				std::string file;
				
				if (!access(argv[i], R_OK))
					file = (!strncmp(argv[i], "/", 1)) ? argv[i] : wd + argv[i];
				else {
					file = listd + argv[i];
					if (access(file.c_str(), R_OK) == -1)
						throw std::runtime_error("can't find " + std::string(argv[i]));
				}
				j->add_file(file.c_str(), target);
			}
		}
		if (!output.first.empty()) {
			range::range_set_ptr rset(new range::range_set_t);
			j->read_files(rset.get());
			list::write_file(output.first.c_str(), output.second, rset.get());
			exit(EXIT_SUCCESS);
		}
	}
	if (client) {
		if (!ps.lock_file(ps.pid_file, F_SETLKW, F_WRLCK, 1, 2))
			throw std::runtime_error("can't lock client range\n");

		if (msgsnd(ps.msqid, j.get(), MSGSIZE(job), 0) == -1)
			throw std::runtime_error("can't send job to msq");

		if (pid_t pid = ps.lock_test(ps.pid_file, F_WRLCK, 0, 1)) {
			if (kill(pid, SIGUSR1) == -1) 
				throw std::runtime_error("can't send signal to iplist\n");
		} else
			throw std::runtime_error("no running iplist instance found");

		if (!ps.lock_file(ps.pid_file, F_SETLK, F_UNLCK, 1, 1))
			throw std::runtime_error("[client] can't unlock request byte\n");

		if (j->req == WRITE)
			list::copy_file(iplist::ps.fifo.c_str(), "/dev/stdout");
		else if (j->req == INSERT && !strncmp(argv[optind], "-", 1))
			list::copy_file("/dev/stdin", iplist::ps.fifo.c_str());

		exit(EXIT_SUCCESS);
	}
	if (msgsnd(ps.msqid, j.get(), MSGSIZE(job), 0) == -1)
		throw std::runtime_error("can't send job to msq");

}

void* iplist::sighandler_start(void*)
{
	int err = EXIT_SUCCESS;
	try {
		int signo;

		for (int ret = 0; !ret;) {
			sigwait(&mask, &signo);

			syslog(LOG_INFO, "info: %s signal caught\n", strsignal(signo));

			pthread_mutex_lock(&lock);
			switch (signo) {
			case SIGINT:
			case SIGQUIT:
			case SIGTERM:
				ret = quit = signo;
			case SIGUSR1:
			case SIGUSR2:
				pthread_cond_signal(&wait);
				break;
			default:
				err++;
			}
			pthread_mutex_unlock(&lock);
		}
		return NULL;

	} catch (const std::exception& e) {
		syslog(LOG_ERR, "thread[%lu]: error: %s\n", pthread_self(), e.what());
		err++;
	}
	exit(err);
}

iplist::job::job():
	mtype(JOB), req(INSERT), file_count(0), nfq_num(NFQ_NUM), 
	policy(POLICY), policy_mark(POLICY_MARK), target_mark(TARGET_MARK),
	range_size(0), ipcount(0), tid(0), pid(0)
{}

void iplist::job::add_file(const char* f, int8_t t) 
{ 
	if (file_count == MAX_FILE) {
		syslog(LOG_WARNING, "warning: too many files, ignoring %s\n", f);
		return;
	}
	strncpy(file[file_count], f, FILE_LENGTH);
	target[file_count++] = t;
}

void iplist::job::add_file(const job& rhs)
{
	for (uint8_t i = 0; i < rhs.file_count; i++)
		this->add_file(rhs.file[i], rhs.target[i]);
}

void iplist::job::read_files(range::range_set_t* rset, bool short_format) const
{
	for (uint8_t i = 0; i < file_count; i++)
		list::read_file(file[i] , rset, target[i], short_format);
}

std::string iplist::job::to_string() const
{
	std::ostringstream os;
	std::map<int8_t, std::string> target_str;

	target_str[-1] = "DEFAULT";
	target_str[NF_DROP] = "DROP";
	target_str[NF_ACCEPT] = "ACCEPT";
	target_str[NF_QUEUE] = "QUEUE";
	target_str[NF_REPEAT] = "REPEAT";

	os << nfq_num << "\t" << target_str[policy] << " (" 
		<< uint32_t(policy_mark) << ")\t" << ipcount << "\t" << range_size << "\t";

	if (file_count > 0)
		os << target_str[target[0]] << " (" << uint32_t(target_mark) 
			<< ")\t" << file[0] << "\n";

	for (uint8_t i = 1; i < file_count; i++)
		os << "\t\t\t\t\t\t" << target_str[target[i]] << " (" 
			<< uint32_t(target_mark) << ")\t" << file[i] << "\n";

	return os.str();
}

